File: unisexBathroom.c
#ifndef _REENTRANT
#define _REENTRANT
#endif

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

#define NUM_MEN 5
#define NUM_WOMEN 5

/* ---------------- Shared State ---------------- */

int men_count = 0;
int women_count = 0;

/* Semaphores */
sem_t mutex;        // protects counters
sem_t bathroom;     // controls bathroom access
sem_t turnstile;    // fairness control

/* ---------------- Thread Functions ---------------- */

void* man_thread(void* arg)
{
    int id = *((int*)arg);

    while (1)
    {
        /* Simulate working outside bathroom */
        sleep(rand() % 3 + 1);

        printf("Man %d wants to enter\n", id);

        /* ===== ENTRY SECTION ===== */
        sem_wait(&mutex);
            men_count++;
            if (men_count == 1)
                sem_wait(&bathroom);
        sem_post(&mutex);

        printf("Man %d enters bathroom\n", id);
        printf("DEBUG: men_count=%d women_count=%d\n", men_count, women_count);


        /* Simulate using bathroom */
        sleep(rand() % 2 + 1);

        printf("DEBUG: men_count=%d women_count=%d\n", men_count, women_count);
        printf("Man %d leaving bathroom\n", id);

        /* ===== EXIT SECTION ===== */
        sem_wait(&mutex);
            men_count--;
            if (men_count == 0)
                sem_post(&bathroom);
        sem_post(&mutex);
    }

    return NULL;
}

void* woman_thread(void* arg)
{
    int id = *((int*)arg);

    while (1)
    {
        sleep(rand() % 3 + 1);

        printf("Woman %d wants to enter\n", id);

        /* ===== ENTRY SECTION ===== */
        sem_wait(&mutex);
            women_count++;
            if (women_count == 1)
                sem_wait(&bathroom);
        sem_post(&mutex);

        printf("Woman %d enters bathroom\n", id);
        printf("DEBUG: men_count=%d women_count=%d\n", men_count, women_count);

        sleep(rand() % 2 + 1);

        printf("DEBUG: men_count=%d women_count=%d\n", men_count, women_count);
        printf("Woman %d leaving bathroom\n", id);

        /* ===== EXIT SECTION ===== */
        sem_wait(&mutex);
            women_count--;
            if (women_count == 0)
                sem_post(&bathroom);
        sem_post(&mutex);
    }

    return NULL;
}

/* ---------------- Main ---------------- */

int main()
{
    pthread_t men[NUM_MEN];
    pthread_t women[NUM_WOMEN];

    int men_ids[NUM_MEN];
    int women_ids[NUM_WOMEN];

    srand(time(NULL));

    /* Initialize semaphores */
    sem_init(&mutex, 0, 1);       // binary semaphore
    sem_init(&bathroom, 0, 1);    // bathroom initially free
    sem_init(&turnstile, 0, 1);   // fairness gate

    /* Create men threads */
    for (int i = 0; i < NUM_MEN; i++)
    {
        men_ids[i] = i;
        pthread_create(&men[i], NULL, man_thread, &men_ids[i]);
    }

    /* Create women threads */
    for (int i = 0; i < NUM_WOMEN; i++)
    {
        women_ids[i] = i;
        pthread_create(&women[i], NULL, woman_thread, &women_ids[i]);
    }

    /* Join threads (program runs forever) */
    for (int i = 0; i < NUM_MEN; i++)
        pthread_join(men[i], NULL);

    for (int i = 0; i < NUM_WOMEN; i++)
        pthread_join(women[i], NULL);

    return 0;
}

File: unisexBathroom.java
import java.util.Random;

enum Turn {
    MEN,
    WOMEN
}
// 
class Bathroom {
    // shared variables/resources between threads need to be protected
    protected int menInside = 0;
    protected int womenInside = 0;
    // unlike rw problem, we implement fairness, which means keeping track
    // of how many threads are currently using the resources and how many are 
    // waiting to use
    protected int waitingMen = 0;
    protected int waitingWomen = 0;

    protected int turn; // 1 for W, 0 for M

    public synchronized void manEnter(){
        waitingMen++;
        while (womenInside > 0 || (turn == 1 && waitingWomen>0)){
            try {
                wait();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); 
            }
        }
        waitingMen--;
        menInside++;
    }

    public synchronized void manExit(){
        menInside--;
        if (menInside == 0) {
            turn = 1;
            notifyAll();
        }
        
    
    }

    public synchronized void womanEnter(){
        waitingWomen++;
        while (menInside > 0 || (turn == 0 && waitingMen>0)){
            try {
                waitingWomen++;
                wait();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); 
            }
        }
        waitingWomen--;
        womenInside++;
    }

    public synchronized void womanExit(){
        womenInside--;
        if(womenInside==0){
            turn = 0;
            notifyAll();
        }
        
    }
}
class Man extends Thread {
    private final int id;
    private final Bathroom bathroom;
    private final Random rand = new Random();

    public Man(int id, Bathroom bathroom) {
        this.id = id;
        this.bathroom = bathroom;
    }

    public void run() {
        while (true) {
            try {
                // Step 1: work outside
                Thread.sleep(rand.nextInt(3000) + 1000);

                // Step 2: request entry
                System.out.println("Man " + id + " wants to enter");
                bathroom.manEnter();
                System.out.println("Man " + id + " enters bathroom");

                // Step 3: use bathroom
                Thread.sleep(rand.nextInt(2000) + 500);

                // Step 4: exit
                System.out.println("Man " + id + " leaving bathroom");
                bathroom.manExit();

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        }
    }
}

class Woman extends Thread {
    private final int id;
    private final Bathroom bathroom;
    private final Random rand = new Random();

    public Woman(int id, Bathroom bathroom) {
        this.id = id;
        this.bathroom = bathroom;
    }

    public void run() {
        while (true) {
            try {
                Thread.sleep(rand.nextInt(3000) + 1000);

                System.out.println("Woman " + id + " wants to enter");
                bathroom.womanEnter();
                System.out.println("Woman " + id + " enters bathroom");

                Thread.sleep(rand.nextInt(2000) + 500);

                System.out.println("Woman " + id + " leaving bathroom");
                bathroom.womanExit();

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        }
    }
}

class unisexBathroom {
    public static void main(String[] args) {
        final int NUM_MEN = 5;
        final int NUM_WOMEN = 5;

        Bathroom bathroom = new Bathroom();  // your monitor class

        // Create men threads
        for (int i = 0; i < NUM_MEN; i++) {
            new Man(i, bathroom).start();
        }

        // Create women threads
        for (int i = 0; i < NUM_WOMEN; i++) {
            new Woman(i, bathroom).start();
        }

        // Threads run forever; no join needed
    }
}
